xcresult/src/xcresult.rs

```diff
@@ -0,0 +1,348 @@
+use indexmap::indexmap;
+use lazy_static::lazy_static;
+use quick_junit::{NonSuccessKind, Report, TestCase, TestCaseStatus, TestSuite, XmlString};
+use std::str;
+use std::{fs, process::Command};
+
+#[derive(Debug, Clone)]
+pub struct XCResult {
+    pub path: String,
+    results_obj: serde_json::Value,
+}
+
+const LEGACY_FLAG_MIN_VERSION: i32 = 70;
+
+fn xcrun<T: AsRef<str>>(args: &[T]) -> anyhow::Result<String> {
+    if !cfg!(target_os = "macos") {
+        return Err(anyhow::anyhow!("xcrun is only available on macOS"));
+    }
+    let mut cmd = Command::new("xcrun");
+    let bin = cmd.args(args.iter().map(|arg| arg.as_ref()));
+    let output = bin.output().map_err(|_| {
+        anyhow::anyhow!("failed to run xcrun -- please make sure you have xcode installed")
+    })?;
+    let result = String::from_utf8(output.stdout)
+        .map_err(|_| anyhow::anyhow!("got non UTF-8 data from xcrun output"))?;
+    Ok(result)
+}
+
+fn xcrun_version() -> anyhow::Result<i32> {
+    let version_raw = xcrun(&["--version"])?;
+    // regex to match version where the output looks like xcrun version 70.
+    let re = regex::Regex::new(r"xcrun version (\d+)")?;
+    lazy_static! {
+        static ref RE: regex::Regex = regex::Regex::new(r"xcrun version (\d+)").unwrap();
+    }
+    let res = re
+        .captures(&version_raw.to_string())
+        .and_then(|capture_group| capture_group.get(1))
+        .map(|version| Ok(version.as_str().parse::<i32>().ok().unwrap_or(0)))
+        .unwrap_or_else(|| Err(anyhow::anyhow!("failed to parse xcrun version")));
+    if let Ok(version) = res {
+        return Ok(version);
+    }
+    Err(anyhow::anyhow!("failed to parse xcrun version"))
+}
+
+fn xcresulttool(
+    path: &str,
+    options: Option<&Vec<&str>>,
+) -> Result<serde_json::Value, anyhow::Error> {
+    let mut base_args = vec!["xcresulttool", "get", "--path", path, "--format", "json"];
+    let version = xcrun_version()?;
+    if version >= LEGACY_FLAG_MIN_VERSION {
+        base_args.push("--legacy");
+    }
+    if let Some(val) = options {
+        base_args.extend(val);
+    }
+    let output = xcrun(&base_args)?;
+    serde_json::from_str(&output)
+        .map_err(|_| anyhow::anyhow!("failed to parse json from xcrun output"))
+}
+
+impl XCResult {
+    pub fn new(path: String) -> Result<XCResult, anyhow::Error> {
+        let binding = fs::canonicalize(path.clone())
```

Comment:
nit: we don't really need ownership of `path`, nor do we need to clone it here—I recommend to double check the other function signatures to see if it's possible to accept `AsRef<str>`

```suggestion
    pub fn new<T: AsRef<str>>(path: T) -> Result<XCResult, anyhow::Error> {
        let binding = fs::canonicalize(path.as_ref())
```